In [ ]:
import pandas as pd



In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/dataset

/content/drive/MyDrive/devign-CodeVul/dataset


In [ ]:
!pip install gdown
!gdown https://drive.google.com/uc?id=1x6hoF7G-tSYxg8AFybggypLZgMGDNHfF


In [ ]:
!python preprocess.py


In [ ]:
train_df = pd.read_json('/content/drive/MyDrive/devign-CodeVul/dataset/train.jsonl', lines=True)


In [ ]:
train_df

project                                 commit_id  target  \
0      FFmpeg  973b1a6b9070e2bf17d17568cbaf4043ce931f51       0   
1      FFmpeg  321b2a9ded0468670b7678b7c098886930ae16b2       0   
2      FFmpeg  5d5de3eba4c7890c2e8077f5b4ae569671d11cf8       0   
3      FFmpeg  57d77b3963ce1023eaf5ada8cba58b9379405cc8       0   
4      FFmpeg  aba232cfa9b193604ed98f3fa505378d006b1b3b       1   
...       ...                                       ...     ...   
21849    qemu  a8170e5e97ad17ca169c64ba87ae2f53850dab4c       0   
21850    qemu  1ea879e5580f63414693655fcf0328559cdce138       0   
21851    qemu  f74990a5d019751c545e9800a3376b6336e77d38       0   
21852    qemu  a89f364ae8740dfc31b321eed9ee454e996dc3c1       0   
21853    qemu  39fb730aed8c5f7b0058845cb9feac0d4b177985       0   

                                                    func    idx  
0      static av_cold int vdadec_init(AVCodecContext ...      0  
1      static int transcode(AVFormatContext **output_...      1  
2      static void v4l2_free_buffer(void *opaque, uin...      2  
3      int av_opencl_buffer_write(cl_mem dst_cl_buf, ...      4  
4      static int r3d_read_rdvo(AVFormatContext *s, A...      5  
...                                                  ...    ...  
21849  static void exynos4210_mct_write(void *opaque,...  27312  
21850  static int no_init_in (HWVoiceIn *hw, audsetti...  27313  
21851  uint32_t HELPER(stfle)(CPUS390XState *env, uin...  27314  
21852  static void pxa2xx_fir_write(void *opaque, hwa...  27315  
21853  static void disas_thumb_insn(CPUARMState *env,...  27316  

[21854 rows x 5 columns]

In [ ]:
test_df =  pd.read_json('/content/drive/MyDrive/devign-CodeVul/dataset/test.jsonl', lines=True)
test_df

project                                 commit_id  target  \
0     FFmpeg  32bf6550cb9cc9f487a6722fe2bfc272a93c1065       0   
1       qemu  c0c24b95542bc1a4dc3fc6ea71475ae04fa69189       1   
2     FFmpeg  5ff998a233d759d0de83ea6f95c383d03d25d88e       1   
3       qemu  24408a7d2b459bed3697367b81ada76518ca96ef       0   
4     FFmpeg  68f593b48433842f3407586679fe07f3e5199ab9       0   
...      ...                                       ...     ...   
2727    qemu  04088adbe0c5adca66adb6022723362ad90ed0fc       0   
2728  FFmpeg  3a25c707fae3c6e99fdda40474c3d74be24cc4c3       0   
2729  FFmpeg  d6604b29ef544793479d7fb4e05ef6622bb3e534       0   
2730    qemu  a8170e5e97ad17ca169c64ba87ae2f53850dab4c       0   
2731  FFmpeg  7104c23bd1a1dcb8a7d9e2c8838c7ce55c30a331       0   

                                                   func    idx  
0     int ff_get_wav_header(AVFormatContext *s, AVIO...      3  
1     static int xen_9pfs_connect(struct XenDevice *...     12  
2     static int subframe_count_exact(FlacEncodeCont...     35  
3     static void ppc_spapr_init(QEMUMachineInitArgs...     43  
4     static int mpeg1_decode_sequence(AVCodecContex...     60  
...                                                 ...    ...  
2727  static void *mpc8544_load_device_tree(target_p...  27266  
2728  static int mov_read_trak(MOVContext *c, ByteIO...  27272  
2729  static av_cold int libschroedinger_encode_init...  27296  
2730  static struct omap_mcbsp_s *omap_mcbsp_init(Me...  27311  
2731  static void rv34_pred_mv(RV34DecContext *r, in...  27317  

[2732 rows x 5 columns]

In [ ]:
valid_df = pd.read_json('/content/drive/MyDrive/devign-CodeVul/dataset/valid.jsonl', lines=True)
valid_df

project                                 commit_id  target  \
0       qemu  aa1530dec499f7525d2ccaa0e3a876dc8089ed1e       1   
1       qemu  21ce148c7ec71ee32834061355a5ecfd1a11f90f       1   
2       qemu  089da572b956ef0f8f5b8d5917358e07892a77c2       1   
3       qemu  98f343395e937fa1db3a28dfb4f303f97cfddd6c       1   
4       qemu  c39ce112b60ffafbaf700853e32bea74cbb2c148       0   
...      ...                                       ...     ...   
2727    qemu  2637c754ccdb286890ed2a8d0d1da775dbd062af       0   
2728    qemu  9012a53f067a78022947e18050b145c34a3dc599       0   
2729    qemu  ad196a9d0c14f681f010bb4b979030ec125ba976       0   
2730    qemu  a6152b52bc50c5cf1cd118a74b483dd3f0748ebd       0   
2731    qemu  196a778428989217b82de042725dc8eb29c8f8d8       1   

                                                   func    idx  
0     static void filter_mirror_setup(NetFilterState...      8  
1     static inline int64_t sub64(const int64_t a, c...     10  
2     void fw_cfg_add_callback(FWCfgState *s, uint16...     15  
3     static void emulated_push_error(EmulatedState ...     22  
4     static void do_busid_cmd(ESPState *s, uint8_t ...     51  
...                                                 ...    ...  
2727  void cpu_dump_state(CPUState *env, FILE *f,\n\...  27264  
2728  static void rtas_start_cpu(PowerPCCPU *cpu_, s...  27269  
2729  void net_slirp_redir(Monitor *mon, const char ...  27275  
2730  int ppc_hash64_handle_mmu_fault(PowerPCCPU *cp...  27276  
2731  static void qxl_reset_surfaces(PCIQXLDevice *d...  27290  

[2732 rows x 5 columns]

In [ ]:
!python /content/drive/MyDrive/devign-CodeVul/evaluator/evaluator.py -a /content/drive/MyDrive/devign-CodeVul/evaluator/tests.jsonl -p /content/drive/MyDrive/devign-CodeVul/evaluator/predictions.txt


/content/drive/MyDrive/devign-CodeVul/evaluator/tests.jsonl
{0: 0, 1: 1, 2: 0, 3: 0, 4: 1}
/content/drive/MyDrive/devign-CodeVul/evaluator/predictions.txt
prediction: {0: 0, 1: 1, 2: 1, 3: 0, 4: 0}
{'Acc': 0.6}


In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/code


/content/drive/MyDrive/devign-CodeVul/code


In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.8/5.8 MB 97.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 50.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 182.4/182.4 KB 25.1 MB/s eta 0:00:00


In [ ]:
!pip install cloud-tpu-client==0.10 https://storage.googleapis.com/tpu-pytorch/wheels/torch_xla-1.8-cp37-cp37m-linux_x86_64.whl
!pip -q install pytorch-lightning==1.2.7 transformers torchmetrics awscli mlflow boto3 pycm


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
ERROR: torch_xla-1.8-cp37-cp37m-linux_x86_64.whl is not a supported wheel on this platform.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 830.7/830.7 KB 12.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 512.4/512.4 KB 23.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.0/4.0 MB 27.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.7/16.7 MB 45.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.7/132.7 KB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.0/63.0 KB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 840.9/840.9 KB 23.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.6/79.6 KB 8.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 86.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
import os
import sys
import logging

import torch


In [ ]:
pip list | grep torch

torch                         1.13.1+cu116
torchaudio                    0.13.1+cu116
torchsummary                  1.5.1
torchtext                     0.14.1
torchvision                   0.14.1+cu116


In [ ]:
!pip -q install pytorch-lightning==1.2.7 transformers torchmetrics awscli mlflow boto3 pycm


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 830.7/830.7 KB 21.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 512.4/512.4 KB 46.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.0/4.0 MB 75.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.7/16.7 MB 74.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.7/132.7 KB 18.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.0/63.0 KB 7.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 840.9/840.9 KB 61.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 269.4/269.4 KB 34.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.3/10.3 MB 89.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.6/79.6 KB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 140.6/140.6 KB 20.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
import torch
torch.cuda.empty_cache()

##CODEBERT RESULTS

In [ ]:
!pip install accelerate
!python run.py \
    --output_dir=./saved_models \
    --model_type=roberta \
    --tokenizer_name=microsoft/codebert-base \
    --model_name_or_path=microsoft/codebert-base \
    --do_train \
    --train_data_file=../dataset/train.jsonl \
    --eval_data_file=../dataset/valid.jsonl \
    --test_data_file=../dataset/test.jsonl \
    --epoch 5 \
    --block_size 400 \
    --train_batch_size 16 \
    --eval_batch_size 32 \
    --learning_rate 2e-5 \
    --max_grad_norm 1.0 \
    --evaluate_during_training \
    --seed 123456  2>&1 | tee train.log

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Some weights of the model checkpoint at microsoft/codebert-base were not used when initializing RobertaForSequenceClassification: ['pooler.dense.bias', 'pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at microsoft/codebert-base and are newly initialized: ['classifier.out_proj.bias', 'classifier.dense.bias', 'classifier.out_proj.we

In [ ]:
!python run.py \
    --output_dir=./saved_models \
    --model_type=roberta \
    --tokenizer_name=microsoft/codebert-base \
    --model_name_or_path=microsoft/codebert-base \
    --do_eval \
    --do_test \
    --train_data_file=../dataset/train.jsonl \
    --eval_data_file=../dataset/valid.jsonl \
    --test_data_file=../dataset/test.jsonl \
    --epoch 5 \
    --block_size 400 \
    --train_batch_size 16 \
    --eval_batch_size 32 \
    --learning_rate 2e-5 \
    --max_grad_norm 1.0 \
    --evaluate_during_training \
    --seed 123456 2>&1 | tee test.log

Some weights of the model checkpoint at microsoft/codebert-base were not used when initializing RobertaForSequenceClassification: ['pooler.dense.bias', 'pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at microsoft/codebert-base and are newly initialized: ['classifier.out_proj.weight', 'classifier.dense.weight', 'classifier.dense.bias', 'classifier.out_proj.bias']
You should probably TRAIN this model on a down-stream task to be 

In [ ]:
!python /content/drive/MyDrive/devign-CodeVul/evaluator/evaluator.py -a /content/drive/MyDrive/devign-CodeVul/dataset/test.jsonl -p ./saved_models/predictions.txt

/content/drive/MyDrive/devign-CodeVul/dataset/test.jsonl
{3: 0, 12: 1, 35: 1, 43: 0, 60: 0, 70: 0, 83: 1, 84: 1, 108: 0, 121: 1, 122: 1, 149: 1, 164: 1, 176: 0, 178: 0, 179: 0, 185: 0, 200: 1, 218: 0, 244: 1, 245: 1, 246: 1, 248: 1, 262: 1, 265: 1, 272: 1, 280: 1, 289: 0, 304: 1, 305: 1, 321: 0, 325: 0, 338: 1, 369: 1, 372: 1, 373: 1, 391: 1, 399: 0, 402: 0, 404: 0, 436: 1, 439: 1, 444: 1, 461: 0, 474: 0, 490: 1, 491: 1, 501: 0, 510: 0, 539: 1, 547: 0, 566: 0, 569: 1, 572: 1, 573: 1, 574: 1, 585: 1, 599: 0, 657: 1, 666: 0, 669: 1, 674: 1, 677: 0, 683: 1, 686: 1, 707: 1, 710: 1, 729: 1, 740: 1, 748: 1, 767: 1, 781: 1, 794: 0, 797: 0, 837: 1, 848: 1, 855: 1, 860: 1, 883: 0, 886: 0, 899: 0, 906: 0, 913: 0, 921: 1, 924: 1, 937: 1, 941: 1, 944: 1, 946: 1, 961: 1, 963: 1, 971: 1, 982: 0, 1035: 1, 1038: 1, 1040: 1, 1046: 1, 1056: 1, 1059: 1, 1060: 0, 1089: 1, 1107: 0, 1108: 0, 1119: 0, 1121: 0, 1122: 0, 1130: 1, 1136: 1, 1137: 1, 1145: 0, 1167: 0, 1194: 0, 1215: 0, 1227: 0, 1277: 1, 1290: 0, 

#CODE2VEC

In [ ]:
train_df.shape

(21854, 5)

In [ ]:
train_df

project                                 commit_id  target  \
0      FFmpeg  973b1a6b9070e2bf17d17568cbaf4043ce931f51       0   
1      FFmpeg  321b2a9ded0468670b7678b7c098886930ae16b2       0   
2      FFmpeg  5d5de3eba4c7890c2e8077f5b4ae569671d11cf8       0   
3      FFmpeg  57d77b3963ce1023eaf5ada8cba58b9379405cc8       0   
4      FFmpeg  aba232cfa9b193604ed98f3fa505378d006b1b3b       1   
...       ...                                       ...     ...   
21849    qemu  a8170e5e97ad17ca169c64ba87ae2f53850dab4c       0   
21850    qemu  1ea879e5580f63414693655fcf0328559cdce138       0   
21851    qemu  f74990a5d019751c545e9800a3376b6336e77d38       0   
21852    qemu  a89f364ae8740dfc31b321eed9ee454e996dc3c1       0   
21853    qemu  39fb730aed8c5f7b0058845cb9feac0d4b177985       0   

                                                    func    idx  
0      static av_cold int vdadec_init(AVCodecContext ...      0  
1      static int transcode(AVFormatContext **output_...      1  
2      static void v4l2_free_buffer(void *opaque, uin...      2  
3      int av_opencl_buffer_write(cl_mem dst_cl_buf, ...      4  
4      static int r3d_read_rdvo(AVFormatContext *s, A...      5  
...                                                  ...    ...  
21849  static void exynos4210_mct_write(void *opaque,...  27312  
21850  static int no_init_in (HWVoiceIn *hw, audsetti...  27313  
21851  uint32_t HELPER(stfle)(CPUS390XState *env, uin...  27314  
21852  static void pxa2xx_fir_write(void *opaque, hwa...  27315  
21853  static void disas_thumb_insn(CPUARMState *env,...  27316  

[21854 rows x 5 columns]

In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/


/content/drive/MyDrive/devign-CodeVul


In [ ]:
!git clone https://github.com/tech-srl/code2vec


fatal: destination path 'code2vec' already exists and is not an empty directory.


In [ ]:
cd code2vec/

/content/drive/MyDrive/devign-CodeVul/code2vec


In [ ]:
!wget https://s3.amazonaws.com/code2vec/model/java14m_model.tar.gz
!tar -xvzf java14m_model.tar.gz

--2023-01-26 23:09:25--  https://s3.amazonaws.com/code2vec/model/java14m_model.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.228.168, 52.216.213.232, 52.216.218.56, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.228.168|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1440921240 (1.3G) [application/x-tar]
Saving to: ‘java14m_model.tar.gz’

java14m_model.tar.g 100%[===================>]   1.34G  34.1MB/s    in 31s     

2023-01-26 23:09:56 (44.8 MB/s) - ‘java14m_model.tar.gz’ saved [1440921240/1440921240]

models/java14_model/saved_model_iter8.release.data-00000-of-00001
models/java14_model/saved_model_iter8.release.index
models/java14_model/saved_model_iter8.release.meta
models/java14_model/dictionaries.bin


In [ ]:
!python --version


Python 3.8.10


In [ ]:
!clang -v
!python -c"import clang.cindex; import inspect; print(inspect.getfile(clang.cindex))"

clang version 10.0.0-4ubuntu1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda-11.2, version 7.0
/usr/local/lib/python3.8/dist-packages/clang/cindex.py


In [ ]:
!awk --version
!curl --version


awk: not an option: --version
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets


In [ ]:
pwd

'/content/drive/MyDrive/devign-CodeVul/code2vec'

In [ ]:
!git clone https://github.com/AmeerHajAli/code2vec_c.git

Cloning into 'code2vec_c'...
remote: Enumerating objects: 26, done.
remote: Counting objects: 100% (26/26), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 26 (delta 3), reused 26 (delta 3), pack-reused 0
Unpacking objects: 100% (26/26), 44.49 KiB | 168.00 KiB/s, done.


In [ ]:
ls

build_extractor.sh  Input.java
CITATION.cff        interactive_predict.py
code2vec_c/         JavaExtractor/
code2vec.py         keras_attention_layer.py
common.py           keras_checkpoint_saver_callback.py
config.py           keras_model.py
CSharpExtractor/    keras_topk_word_predictions_layer.py
extractor.py        LICENSE
images/             README.md
__init__.py


In [ ]:
cd code2vec_c/

[Errno 2] No such file or directory: 'code2vec_c/'
/content/drive/MyDrive/devign-CodeVul/code2vec/code2vec_c


In [ ]:
pwd

'/content/drive/MyDrive/devign-CodeVul/code2vec/code2vec_c'

In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/code2vec-code

/content/drive/MyDrive/devign-CodeVul/code2vec-code


In [ ]:
!pip install tensorflow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/code2vec-code/astminer


/content/drive/MyDrive/devign-CodeVul/code2vec-code/astminer


In [ ]:
!./gradlew shadowJar


UsageError: Cell magic `%%!./gradlew` not found.


In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec


/content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec


In [ ]:
!pip install dos2unix

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for dos2unix: filename=dos2unix-1-py3-none-any.whl size=1515 sha256=d6e91352b2a280ef1294ced7188dd17c38e453fd5bca91ddfc95e51dcb50f7ca
  Stored in directory: /root/.cache/pip/wheels/9b/51/61/ed824be933dc079a669a77af99b90808edd35dbfb4b5bd22bf
Successfully built dos2unix


In [ ]:
cd /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/

[Errno 2] No such file or directory: '/content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/'
/content


In [ ]:
!source preprocess.sh


Creating histograms from the training data
2023-02-16 23:49:21.957831: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-16 23:49:22.833640: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/lib64-nvidia
2023-02-16 23:49:22.833752: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/lib64-nvidia
2023-02-16 23:49:22.833773: W tensorflow/co

In [ ]:
ls

 code2vec.py            images/                  prediction_outputter.py
 common.py              __init__.py              preprocess.py
 config.py              input.js                 preprocess.sh
 data/                  interactive_predict.py   __pycache__/
'data'$'\r'/            LICENSE                  README.md
 devign.test.raw.txt    log.txt                  requirements.txt
 devign.train.raw.txt   model_base.py            tensorflow_model.py
 devign.valid.raw.txt   models/                  train.sh
 extractor_js.py        path_context_reader.py   vocabularies.py


In [ ]:
!source train.sh

2023-02-06 18:20:50,450 INFO     ---------------------------------------------------------------------
2023-02-06 18:20:50,450 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-06 18:20:50,450 INFO     ---------------------------------------------------------------------
2023-02-06 18:20:50,450 INFO     Checking number of examples ...
2023-02-06 18:20:50,459 INFO         Number of train examples: 237
2023-02-06 18:20:50,459 INFO     Path: data/devign/devign.val.c2v
2023-02-06 18:20:50,472 INFO         Number of test examples: 520
2023-02-06 18:20:50,472 INFO     ---------------------------------------------------------------------
2023-02-06 18:20:50,472 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-06 18:20:50,472 INFO     CODE_VECTOR_SIZE                          384
2023-02-06 18:20:50,472 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-06 18:20:50,473 INFO     DEFAULT_EMBEDDINGS_SIZE 

In [ ]:
 !python3 code2vec.py --load models/devign/saved_model_iter5 --release

2023-02-06 18:22:07,994 INFO     ---------------------------------------------------------------------
2023-02-06 18:22:07,994 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-06 18:22:07,994 INFO     ---------------------------------------------------------------------
2023-02-06 18:22:07,994 INFO     ---------------------------------------------------------------------
2023-02-06 18:22:07,994 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-06 18:22:07,995 INFO     CODE_VECTOR_SIZE                          384
2023-02-06 18:22:07,995 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-06 18:22:07,995 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-06 18:22:07,995 INFO     DL_FRAMEWORK                              tensorflow
2023-02-06 18:22:07,995 INFO     DROPOUT_KEEP_RATE                         0.75
2023-02-06 18:22:07,995 INFO     EXPORT_CODE_VECTORS                       F

In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --test data/devign/devign.test.c2v

2023-02-06 18:23:18,222 INFO     ---------------------------------------------------------------------
2023-02-06 18:23:18,222 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-06 18:23:18,222 INFO     ---------------------------------------------------------------------
2023-02-06 18:23:18,222 INFO     Checking number of examples ...
2023-02-06 18:23:18,222 INFO     Path: data/devign/devign.test.c2v
2023-02-06 18:23:18,224 INFO         Number of test examples: 359
2023-02-06 18:23:18,224 INFO     ---------------------------------------------------------------------
2023-02-06 18:23:18,224 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-06 18:23:18,224 INFO     CODE_VECTOR_SIZE                          384
2023-02-06 18:23:18,224 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-06 18:23:18,224 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-06 18:23:18,224 INFO     DL_FRAMEWOR

#Precision: 0.5925925925925926, Sensitivity/Recall: 0.7804878048780488, Accuracy: 0.5658263305322129, Error Rate: 0.4341736694677871, F1: 0.6736842105263158, #TPs=160, #TNs=42, #FPs=110, #FNs=45, TNR=0.27631578947368424, FPR=0.7236842105263158


In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --save_w2v models/embeddings/tokens.txt


2023-02-06 18:26:11,192 INFO     ---------------------------------------------------------------------
2023-02-06 18:26:11,193 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-06 18:26:11,193 INFO     ---------------------------------------------------------------------
2023-02-06 18:26:11,193 INFO     Checking number of examples ...
2023-02-06 18:26:11,193 INFO     ---------------------------------------------------------------------
2023-02-06 18:26:11,193 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-06 18:26:11,193 INFO     CODE_VECTOR_SIZE                          384
2023-02-06 18:26:11,193 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-06 18:26:11,193 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-06 18:26:11,193 INFO     DL_FRAMEWORK                              tensorflow
2023-02-06 18:26:11,193 INFO     DROPOUT_KEEP_RATE                         0.75
2023-02-06 

In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --save_t2v models/embeddings/targets.txt

2023-02-06 18:28:47,522 INFO     ---------------------------------------------------------------------
2023-02-06 18:28:47,522 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-06 18:28:47,523 INFO     ---------------------------------------------------------------------
2023-02-06 18:28:47,523 INFO     Checking number of examples ...
2023-02-06 18:28:47,523 INFO     ---------------------------------------------------------------------
2023-02-06 18:28:47,523 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-06 18:28:47,523 INFO     CODE_VECTOR_SIZE                          384
2023-02-06 18:28:47,523 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-06 18:28:47,523 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-06 18:28:47,523 INFO     DL_FRAMEWORK                              tensorflow
2023-02-06 18:28:47,523 INFO     DROPOUT_KEEP_RATE                         0.75
2023-02-06 

In [ ]:
!python code2vec.py --load models/devign/saved_model_iter5.release --predict --export_code_vectors

2023-02-08 17:52:38,337 INFO     ---------------------------------------------------------------------
2023-02-08 17:52:38,337 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 17:52:38,337 INFO     ---------------------------------------------------------------------
2023-02-08 17:52:38,337 INFO     Checking number of examples ...
2023-02-08 17:52:38,337 INFO     ---------------------------------------------------------------------
2023-02-08 17:52:38,337 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 17:52:38,337 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 17:52:38,337 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 17:52:38,337 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-08 17:52:38,337 INFO     DL_FRAMEWORK                              tensorflow
2023-02-08 17:52:38,337 INFO     DROPOUT_KEEP_RATE                         0.75
2023-02-08 

In [ ]:
!python prediction_outputter.py --load models/devign/saved_model_iter5.release --predict --export_code_vectors
# or your chosen iteration
# !python /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/prediction_outputter.py -a /content/drive/MyDrive/devign-CodeVul/code2vec-code/astminer/dataset/test.jsonl -p models/devign/saved_model_iter5.release 

2023-02-08 17:13:14.621146: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:42] Overriding orig_value setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.
2023-02-08 17:13:14,622 INFO     ---------------------------------------------------------------------
2023-02-08 17:13:14,622 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 17:13:14,622 INFO     ---------------------------------------------------------------------
2023-02-08 17:13:14,622 INFO     Checking number of examples ...
2023-02-08 17:13:14,622 INFO     ---------------------------------------------------------------------
2023-02-08 17:13:14,622 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 17:13:14,622 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 17:13:14,622 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 17:13:14,622 INFO     DEFAULT

In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --export_code_vectors --test /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.test.c2v



2023-02-08 18:01:50,473 INFO     ---------------------------------------------------------------------
2023-02-08 18:01:50,473 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 18:01:50,473 INFO     ---------------------------------------------------------------------
2023-02-08 18:01:50,473 INFO     Checking number of examples ...
2023-02-08 18:01:50,473 INFO     Path: /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.test.c2v
2023-02-08 18:01:51,247 INFO         Number of test examples: 359
2023-02-08 18:01:51,247 INFO     ---------------------------------------------------------------------
2023-02-08 18:01:51,247 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 18:01:51,247 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 18:01:51,247 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 18:01:51,247 INFO     DEFAULT_EMBEDDINGS_SIZE      

###2023-02-08 18:01:53,984 INFO     Precision: 0.5925925925925926, Sensitivity/Recall: 0.7804878048780488, Accuracy: 0.5658263305322129, Error Rate: 0.4341736694677871, F1: 0.6736842105263158, #TPs=160, #TNs=42, #FPs=110, #FNs=45, TNR=0.27631578947368424, FPR=0.7236842105263158



In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --test /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.train.c2v --export_code_vectors

2023-02-08 18:48:46,829 INFO     ---------------------------------------------------------------------
2023-02-08 18:48:46,829 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 18:48:46,829 INFO     ---------------------------------------------------------------------
2023-02-08 18:48:46,829 INFO     Checking number of examples ...
2023-02-08 18:48:46,829 INFO     Path: /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.train.c2v
2023-02-08 18:48:46,830 INFO         Number of test examples: 237
2023-02-08 18:48:46,830 INFO     ---------------------------------------------------------------------
2023-02-08 18:48:46,830 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 18:48:46,831 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 18:48:46,831 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 18:48:46,831 INFO     DEFAULT_EMBEDDINGS_SIZE     

2023-02-08 18:48:49,469 INFO     Precision: 0.7720588235294118, Sensitivity/Recall: 0.9210526315789473, Accuracy: 0.8312236286919831, Error Rate: 0.16877637130801687, F1: 0.8400000000000001, #TPs=105, #TNs=92, #FPs=31, #FNs=9, TNR=0.7479674796747967, FPR=0.25203252032520324


In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --test /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.val.c2v --export_code_vectors

2023-02-08 18:49:52,516 INFO     ---------------------------------------------------------------------
2023-02-08 18:49:52,517 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 18:49:52,517 INFO     ---------------------------------------------------------------------
2023-02-08 18:49:52,517 INFO     Checking number of examples ...
2023-02-08 18:49:52,517 INFO     Path: /content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.val.c2v
2023-02-08 18:49:52,518 INFO         Number of test examples: 520
2023-02-08 18:49:52,518 INFO     ---------------------------------------------------------------------
2023-02-08 18:49:52,518 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 18:49:52,519 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 18:49:52,519 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 18:49:52,519 INFO     DEFAULT_EMBEDDINGS_SIZE       

2023-02-08 18:49:55,281 INFO     Precision: 0.44554455445544555, Sensitivity/Recall: 0.7627118644067796, Accuracy: 0.4605009633911368, Error Rate: 0.5394990366088632, F1: 0.5625, #TPs=180, #TNs=59, #FPs=224, #FNs=56, TNR=0.20848056537102475, FPR=0.7915194346289752


In [ ]:
!python3 code2vec.py --load models/devign/saved_model_iter5.release --export_code_vectors --predict


2023-02-08 18:23:06,617 INFO     ---------------------------------------------------------------------
2023-02-08 18:23:06,617 INFO     ---------------------- Creating code2vec model ----------------------
2023-02-08 18:23:06,617 INFO     ---------------------------------------------------------------------
2023-02-08 18:23:06,617 INFO     Checking number of examples ...
2023-02-08 18:23:06,617 INFO     ---------------------------------------------------------------------
2023-02-08 18:23:06,617 INFO     ----------------- Configuration - Hyper Parameters ------------------
2023-02-08 18:23:06,617 INFO     CODE_VECTOR_SIZE                          384
2023-02-08 18:23:06,617 INFO     CSV_BUFFER_SIZE                           104857600
2023-02-08 18:23:06,617 INFO     DEFAULT_EMBEDDINGS_SIZE                   128
2023-02-08 18:23:06,617 INFO     DL_FRAMEWORK                              tensorflow
2023-02-08 18:23:06,617 INFO     DROPOUT_KEEP_RATE                         0.75
2023-02-08 

In [ ]:
with open('/content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/data/devign/devign.test.c2v.vectors','r') as firstfile, open('/content/drive/MyDrive/devign-CodeVul/code2vec-code/vector_results.txt','w') as secondfile:
    for line in firstfile:
             secondfile.write(line)

In [ ]:
import pandas as pd
vectors_table = pd.read_csv('/content/drive/MyDrive/devign-CodeVul/code2vec-code/vector_results.txt', sep=" ", header=None)

In [ ]:
vectors_table

0         1         2         3         4         5         6    \
0    0.027397 -0.019585 -0.013053 -0.019990 -0.000280  0.018346  0.010392   
1   -0.022493  0.034752 -0.019768 -0.008014 -0.032480  0.008879  0.103854   
2    0.024675  0.043641 -0.044374 -0.045136 -0.018894 -0.011160  0.073667   
3   -0.019524  0.026950 -0.011974 -0.013604 -0.012386  0.006614  0.041252   
4   -0.022803  0.041887 -0.025068 -0.030957 -0.021506  0.001983  0.010321   
..        ...       ...       ...       ...       ...       ...       ...   
352 -0.009736  0.047279  0.005255  0.013753 -0.008843  0.007202  0.074640   
353 -0.018987  0.020552 -0.029790 -0.029172  0.004728  0.002650  0.050022   
354  0.018854  0.040439 -0.029264 -0.021056 -0.032004  0.015466  0.065553   
355  0.007958 -0.002836 -0.009997 -0.022277 -0.026563  0.003658  0.090518   
356  0.013366  0.052146 -0.008595 -0.045030 -0.001026 -0.027820  0.019919   

          7         8         9    ...       374       375       376  \
0    0.035380 -0.008344  0.075914  ... -0.027436 -0.018418  0.004052   
1    0.022706 -0.060262  0.036250  ... -0.058376 -0.032297 -0.035698   
2    0.031929 -0.047267  0.019654  ... -0.015139 -0.022987 -0.015272   
3    0.014665 -0.032114  0.029717  ... -0.055452  0.012167 -0.041464   
4    0.039018 -0.059125 -0.002012  ... -0.020411 -0.026096 -0.056039   
..        ...       ...       ...  ...       ...       ...       ...   
352  0.039383 -0.080710  0.029368  ... -0.083585  0.011601 -0.027306   
353  0.032604 -0.046397  0.029236  ... -0.020794 -0.024058  0.001452   
354  0.035427 -0.029733  0.024650  ... -0.045160 -0.034835 -0.039974   
355  0.020016 -0.045624  0.037074  ... -0.058789 -0.009086 -0.015466   
356  0.036319 -0.026608  0.033969  ... -0.044899 -0.062364 -0.036412   

          377       378       379       380       381       382       383  
0   -0.014924 -0.022526 -0.017346 -0.043375 -0.015443 -0.010276  0.063321  
1   -0.029021  0.026015  0.033951  0.020788  0.018774 -0.042694  0.011068  
2   -0.012139 -0.020569  0.042732 -0.015291 -0.015469 -0.060082  0.040120  
3   -0.032086  0.016695 -0.024321  0.030963  0.009650 -0.007788  0.016668  
4    0.005714 -0.005483  0.002489  0.004314 -0.001680 -0.064337  0.023834  
..        ...       ...       ...       ...       ...       ...       ...  
352 -0.014982  0.021017  0.034008  0.005772  0.028297 -0.019645 -0.014497  
353 -0.019228 -0.039378  0.026591 -0.044416  0.015559 -0.033816  0.024650  
354  0.002712 -0.017800  0.049537  0.000398 -0.014715 -0.046561  0.003249  
355 -0.031261  0.009568  0.031238  0.017694  0.043526 -0.034827 -0.000089  
356 -0.010349  0.002528  0.035826 -0.032699 -0.005173 -0.053789  0.026182  

[357 rows x 384 columns]

In [ ]:
vectors_table = pd.read_csv('/content/drive/MyDrive/devign-CodeVul/code2vec-code/code2vec/devign.train.raw.txt', sep=" ", header=None)


In [ ]:
vectors_table

0                                                  1    \
0    safe  av|codec|context|*,c4b44d57c510ae50e3f1f4c368b...   
1    safe  av|format|context|*|*,c4b44d57c510ae50e3f1f4c3...   
2    safe     void|*,c4b44d57c510ae50e3f1f4c368b9e232,opaque   
3    safe  cl|mem,c4b44d57c510ae50e3f1f4c368b9e232,dst|cl...   
4    vuln  av|format|context|*,c4b44d57c510ae50e3f1f4c368...   
..    ...                                                ...   
232  safe  data,ef0f501e85af13c3f3106da4bf2906e0,test|out...   
233  safe  s,ef0f501e85af13c3f3106da4bf2906e0,ac|encode|c...   
234  safe     sd|state|*,c4b44d57c510ae50e3f1f4c368b9e232,sd   
235  safe  e,ef0f501e85af13c3f3106da4bf2906e0,event|notif...   
236  safe       visitor|*,c4b44d57c510ae50e3f1f4c368b9e232,v   

                                                   2    \
0    ctx,7287fb7dd01a251fda0dbd3ae74001c1,vda|decod...   
1    int,c4b44d57c510ae50e3f1f4c368b9e232,nb|output...   
2     unused,ef0f501e85af13c3f3106da4bf2906e0,uint|t|*   
3    uint|t|*,c4b44d57c510ae50e3f1f4c368b9e232,src|buf   
4         atom,ef0f501e85af13c3f3106da4bf2906e0,atom|*   
..                                                 ...   
232  const|void|*,c4b44d57c510ae50e3f1f4c368b9e232,...   
233           blk,7287fb7dd01a251fda0dbd3ae74001c1,int   
234      uint|t,c4b44d57c510ae50e3f1f4c368b9e232,value   
235  e,06f1bd4eabc1f73da38b772edea29dca,event|notif...   
236  name,ef0f501e85af13c3f3106da4bf2906e0,const|ch...   

                                                   3    \
0    ctx,06f1bd4eabc1f73da38b772edea29dca,vda|decod...   
1    input|file|*,c4b44d57c510ae50e3f1f4c368b9e232,...   
2    v|l|buffer|*,8a2b9543a7267ced6831ca9e368e2149,...   
3     size|t,c4b44d57c510ae50e3f1f4c368b9e232,buf|size   
4    r|d,7287fb7dd01a251fda0dbd3ae74001c1,r|d|conte...   
..                                                 ...   
232  test|output|visitor|data|*,be3b43d7f4a43891e03...   
233            ch,7287fb7dd01a251fda0dbd3ae74001c1,int   
234             int,8a2b9543a7267ced6831ca9e368e2149,i   
235              e,4f27c208767e095fd30422fc396d9a65,fd   
236      uint|t|*,c4b44d57c510ae50e3f1f4c368b9e232,obj   

                                                   4    \
0    av|codec|context|*,be3b43d7f4a43891e0385f2eca2...   
1    nb|input|files,ef0f501e85af13c3f3106da4bf2906e...   
2    avbuf,06f1bd4eabc1f73da38b772edea29dca,v|l|buf...   
3       cl|int,8a2b9543a7267ced6831ca9e368e2149,status   
4    r|d|context|*,be3b43d7f4a43891e0385f2eca2cf07f...   
..                                                 ...   
232  test|output|visitor|data|*,d5ca1a6cadb392e0831...   
233             i,7287fb7dd01a251fda0dbd3ae74001c1,int   
234     sd|state|*,be3b43d7f4a43891e0385f2eca2cf07f,sd   
235  event|notifier|*,d5ca1a6cadb392e083136474a8cab...   
236    error|*|*,c4b44d57c510ae50e3f1f4c368b9e232,errp   

                                                   5    \
0    av|codec|context|*,d5ca1a6cadb392e083136474a8c...   
1    stream|map|*,c4b44d57c510ae50e3f1f4c368b9e232,...   
2       void|*,be3b43d7f4a43891e0385f2eca2cf07f,opaque   
3       mapped,7287fb7dd01a251fda0dbd3ae74001c1,void|*   
4    s,06f1bd4eabc1f73da38b772edea29dca,av|format|c...   
..                                                 ...   
232          data,4f27c208767e095fd30422fc396d9a65,qov   
233    group|size,7287fb7dd01a251fda0dbd3ae74001c1,int   
234   sd|state|*,d5ca1a6cadb392e083136474a8cabc30,bdrv   
235  event|notifier|get|fd,983bf92b87bd6bebbf29a788...   
236  opts|visitor|*,8a2b9543a7267ced6831ca9e368e214...   

                                                   6    \
0     avctx,4f27c208767e095fd30422fc396d9a65,priv|data   
1    int,c4b44d57c510ae50e3f1f4c368b9e232,nb|stream...   
2        avbuf,fb9f9251d9b23ebb99f316967367cfc9,void|*   
3       mapped,06f1bd4eabc1f73da38b772edea29dca,void|*   
4         s,4f27c208767e095fd30422fc396d9a65,priv|data   
..                                                 ...   
232  test|output|visitor|data|*,052c0d4821e

In [ ]:
result = pd.read_csv('/content/drive/MyDrive/devign-CodeVul/code2vec-code/merge_all.vectors', sep=" ", header=None)


In [ ]:
result

0         1         2         3         4         5         6    \
0     0.027397 -0.019585 -0.013053 -0.019990 -0.000280  0.018346  0.010392   
1    -0.022493  0.034752 -0.019768 -0.008014 -0.032480  0.008879  0.103854   
2     0.024675  0.043641 -0.044374 -0.045136 -0.018894 -0.011160  0.073667   
3    -0.019524  0.026950 -0.011974 -0.013604 -0.012386  0.006614  0.041252   
4    -0.022803  0.041887 -0.025068 -0.030957 -0.021506  0.001983  0.010321   
...        ...       ...       ...       ...       ...       ...       ...   
1108  0.053500  0.035493 -0.017215 -0.009382 -0.033096  0.004248  0.086625   
1109 -0.028406  0.025998 -0.053398 -0.014485 -0.011894 -0.019773  0.057710   
1110  0.008020  0.058083 -0.049507 -0.021078 -0.011355  0.045467  0.068319   
1111  0.024923  0.058456 -0.021010  0.016159 -0.010869  0.008751  0.084613   
1112  0.021734 -0.032089 -0.016246 -0.000460 -0.086942 -0.042308  0.014972   

           7         8         9    ...       375       376       377  \
0     0.035380 -0.008344  0.075914  ... -0.018418  0.004052 -0.014924   
1     0.022706 -0.060262  0.036250  ... -0.032297 -0.035698 -0.029021   
2     0.031929 -0.047267  0.019654  ... -0.022987 -0.015272 -0.012139   
3     0.014665 -0.032114  0.029717  ...  0.012167 -0.041464 -0.032086   
4     0.039018 -0.059125 -0.002012  ... -0.026096 -0.056039  0.005714   
...        ...       ...       ...  ...       ...       ...       ...   
1108  0.040455 -0.034407  0.025079  ... -0.054857 -0.024456 -0.019349   
1109  0.012316 -0.056061  0.030213  ... -0.010136 -0.033195 -0.010198   
1110  0.007506 -0.042748  0.013853  ... -0.015225 -0.018784 -0.022156   
1111  0.021934  0.004572  0.011758  ... -0.014669 -0.028630 -0.032371   
1112  0.054032 -0.065831  0.071117  ...  0.032753 -0.038515 -0.029017   

           378       379       380       381       382       383  384  
0    -0.022526 -0.017346 -0.043375 -0.015443 -0.010276  0.063321    1  
1     0.026015  0.033951  0.020788  0.018774 -0.042694  0.011068    0  
2    -0.020569  0.042732 -0.015291 -0.015469 -0.060082  0.040120    0  
3     0.016695 -0.024321  0.030963  0.009650 -0.007788  0.016668    1  
4    -0.005483  0.002489  0.004314 -0.001680 -0.064337  0.023834    1  
...        ...       ...       ...       ...       ...       ...  ...  
1108 -0.014026  0.019855 -0.045793 -0.014315 -0.049802  0.039327    1  
1109  0.021070  0.034495  0.010973  0.009258 -0.047495  0.014458    1  
1110 -0.002408  0.040017  0.020709  0.002150 -0.025339  0.003485    1  
1111  0.025256  0.052429  0.008627  0.001961 -0.029948  0.013538    1  
1112  0.008190  0.009181 -0.042520  0.079037 -0.002422  0.022014    1  

[1113 rows x 385 columns]

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import datasets
from sklearn.model_selection import cross_val_score
from gensim.models import word2vec, FastText
import re
from sklearn.metrics import ( accuracy_score, roc_auc_score,
                             precision_score, recall_score,
                             classification_report
                             )
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
import plotly.graph_objects as go
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.metrics import classification_report
from sklearn.cluster import AgglomerativeClustering
import scipy.cluster.hierarchy as shc
from sklearn.cluster import DBSCAN
from sklearn.cluster import SpectralClustering
from sklearn.mixture import GaussianMixture
from sklearn import cluster, datasets, mixture
from sklearn import metrics
from numpy import unique
from numpy import where
from matplotlib import pyplot
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV

from sklearn.inspection import permutation_importance
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

from sklearn.utils import shuffle
code = shuffle(result)
print(code.head(5))

X = code.iloc[:,:].values

z=pd.DataFrame(X)
z

X=z.loc[:,:383]
y=code.iloc[:,-1]
print(y)

scaler = StandardScaler()
scaled = scaler.fit_transform(X)
print(scaled)
X = pd.DataFrame(scaled)
print(X)


          0         1         2         3         4         5         6    \
189  0.002040  0.001650 -0.022966  0.004354 -0.019843 -0.003627  0.065163   
153  0.042945  0.015525 -0.028666 -0.003570  0.011589 -0.025428  0.047149   
349  0.053688  0.051505 -0.026024 -0.039041 -0.034274  0.021735  0.048307   
479  0.025126  0.011205 -0.017889  0.001224 -0.018747  0.011061 -0.019184   
319  0.049926  0.040590 -0.026192 -0.024015 -0.013040 -0.029761  0.007440   

          7         8         9    ...       375       376       377  \
189  0.016677 -0.016221  0.015496  ... -0.047699 -0.037720 -0.015719   
153  0.029694  0.013272  0.023179  ... -0.019071 -0.034846 -0.033863   
349  0.061047 -0.048755  0.036706  ... -0.051918 -0.066359 -0.030936   
479 -0.002597  0.027240  0.012962  ... -0.060000 -0.067572 -0.016253   
319  0.063102  0.005340  0.022210  ... -0.040948 -0.024904 -0.011957   

          378       379       380       381       382       383  384  
189  0.031102  0.026385  0.012570

In [ ]:
X.shape

(1113, 384)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3) # 70% training and 30% test

clf = RandomForestClassifier(n_estimators=100)

clf.fit(X_train,y_train)

y_pred=clf.predict(X_test)

print("Accuracy:",metrics.accuracy_score(y_test, y_pred))


prec_rec = classification_report(y_pred, y_test)
print("CLASSIFICATION REPORT")
print(prec_rec)
print("precision", metrics.average_precision_score(y_test, y_pred))
print("recall", metrics.recall_score(y_test, y_pred))
print("f1", metrics.f1_score(y_test, y_pred))
print("f2", metrics.fbeta_score(y_test,y_pred, beta=2))

Accuracy: 0.5598802395209581
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.59      0.55      0.57       176
           1       0.53      0.57      0.55       158

    accuracy                           0.56       334
   macro avg       0.56      0.56      0.56       334
weighted avg       0.56      0.56      0.56       334

precision 0.539875009923229
recall 0.5325443786982249
f1 0.5504587155963304
f2 0.539568345323741


In [ ]:
#IMPROVED RF


rfc=RandomForestClassifier(random_state=42)
param_grid = { 
    'n_estimators': [200, 500],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth' : [8, 16, 32],
    'criterion' :['gini', 'entropy']
}
CV_rfc = GridSearchCV(estimator=rfc, param_grid=param_grid, cv= 5)
CV_rfc.fit(X_train, y_train)
CV_rfc.best_params_

{'criterion': 'entropy',
 'max_depth': 8,
 'max_features': 'log2',
 'n_estimators': 500}

In [ ]:
rfc1=RandomForestClassifier(random_state=42, max_features='log2', n_estimators= 200, max_depth=8, criterion ='entropy')
rfc1.fit(X, y)
pred=rfc1.predict(X_test)
print("Accuracy for Random Forest on CV data: ",accuracy_score(y_test,pred))

prec_rec = classification_report(pred, y_test)
print("CLASSIFICATION REPORT")
print(prec_rec)
print("precision", metrics.average_precision_score(y_test, pred))
print("recall", metrics.recall_score(y_test, pred))
print("f1", metrics.f1_score(y_test, pred))
print("f2", metrics.fbeta_score(y_test,pred, beta=2))

Accuracy for Random Forest on CV data:  0.9880239520958084
CLASSIFICATION REPORT
              precision    recall  f1-score   support

           0       0.99      0.98      0.99       167
           1       0.98      0.99      0.99       167

    accuracy                           0.99       334
   macro avg       0.99      0.99      0.99       334
weighted avg       0.99      0.99      0.99       334

precision 0.9853488289692803
recall 0.9822485207100592
f1 0.988095238095238
f2 0.9845788849347569


In [ ]:
#Pruning 

from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.metrics import classification_report
from sklearn.cluster import AgglomerativeClustering
import scipy.cluster.hierarchy as shc
from sklearn.cluster import DBSCAN
from sklearn.cluster import SpectralClustering
from sklearn.mixture import GaussianMixture
from sklearn import cluster, datasets, mixture
from sklearn import metrics
from numpy import unique
from numpy import where
from matplotlib import pyplot
from sklearn.model_selection import cross_val_score
import matplotlib.pyplot as plt


In [ ]:
class PrunedRandomForest:


    def __init__(self, n_clusters = 10 , n_estimators = 500, cv = 5):
        """ this is the unpruned random forest """
        self.rf = RandomForestClassifier(n_estimators = n_estimators)

        """ this is the final pruned random forest """
        self.prf = None 
        self.prf1 = None 

        
        """this is a list of decision tree object present in original unpruned random forest"""
        self.decision_trees = None 
        
        """number of clusters for pruning"""
        self.n_clusters = n_clusters 
        
        

        """ 
            this contains dictionary of lists where each element of dictionary represetns a
            pair (cluster_idx, tree_idx_list)

            where tree_idx_list is list of decision tree indices and cluster_idx is index of cluster 
            e.g.
            
            {
                0 : [2, 0]
                1 : [1, 3]
            }
            it means decision tree at 2nd index and 0th index 
            are in same cluster similarly 1st and 3rd in second cluster

        """ 
        self.clusters = {}
        
        self.cv = cv


        """
            Information about all trees in the random forest present in this 
            form
            {
                idx : { 
                    accuracy : 89.34
                    cluster : 2
                }

                idx1 : {

                }
            }
        """
        self.info = {}
        #self.maxs_accuracy = {}
        self.fimps = None

    def prune(self):
        pass

    
    def fit(self, X, y):
        self.rf.fit(X,y)
        self.decision_trees = self.rf.estimators_
        self.create_feature_imp_clusters()
        self.create_feature_imp_clusters1()
        self.create_fimp_agglomerative()
        self.create_fimp_gaussian()
        self.create_fimp_spectral_clustering()
        #self.plot_clusters(X,y)
        #self.all_algorithms()

        self.fill_info(X, y)
        #self.max_accuracy(X,y)
        self.prune()

    def fill_info(self, X, y):
        for idx, tree in enumerate(self.decision_trees):
            self.info[idx] = self.fill_for_tree(tree, X, y)

        for cluster_idx in self.clusters:
            for tree_idx in self.clusters[cluster_idx]:
                self.info[tree_idx]["cluster_idx"] = cluster_idx 
       

    """
        {
            "accuracy" : 98.45,
            
        } 
    """


    def fill_for_tree(self, tree, X, y):
        info = {}
        mean_score = cross_val_score(tree, X, y, cv=self.cv).mean()
        info["accuracy"] = (mean_score)
        return info

    def predict(self, X):
        if(self.prf is None):
            return None
        
        return self.prf.predict(X)

    
    def get_decision_trees(self):
        return self.decision_trees

    def get_feature_imps_from_rf(self):
        feature_imp_list = []
        dts = self.get_decision_trees()
        for tree in dts:
            feature_imp_list.append(tree.feature_importances_)
        return feature_imp_list

    def create_feature_imp_clusters(self):
        fimps = self.get_feature_imps_from_rf()
        #print(fimps)
        #print("KMEANS")
        kmeans = KMeans(n_clusters=self.n_clusters, random_state=0).fit(fimps)

        print(" fimps len = " + str(len(fimps)))
        labels = kmeans.labels_
        score = silhouette_score(fimps, labels, metric='euclidean')
        print('Silhouetter Score: %.3f' % score)
        #print(labels)
        print(" kmeans.labels_ len = " + str(len(kmeans.labels_)))
        for (tree_idx, cluster_idx) in enumerate(kmeans.labels_):
            if cluster_idx in self.clusters:
                self.clusters[cluster_idx].append(tree_idx)
            else:
                self.clusters[cluster_idx] = [tree_idx]
        
    def create_feature_imp_clusters1(self):
        fimps = self.get_feature_imps_from_rf()
        #print("DBSCAN-")
        dbscan = DBSCAN().fit(fimps)

        print(" fimps len = " + str(len(fimps)))
        labels = dbscan.labels_
        #print(labels)
        print(" dbscan.labels_ len = " + str(len(labels)))
        

    def create_fimp_agglomerative(self):
        fimps = self.get_feature_imps_from_rf()
        #print("AGGLOMERATIVE CLUSTERING-")
        groups = AgglomerativeClustering(n_clusters=self.n_clusters).fit(fimps)
        #print(groups.labels_)

    def create_fimp_gaussian(self):
      fimps = self.get_feature_imps_from_rf()
      #print("GAUSSIAN CLUSTERING-")
      gmm =GaussianMixture(n_components = 5)
      gmm.fit(fimps)
      labelsgmm = gmm.predict(fimps)
      #print(labelsgmm)
      
    def create_fimp_spectral_clustering(self):
      fimps = self.get_feature_imps_from_rf()
      models = SpectralClustering(n_clusters=6, random_state=25, n_neighbors=8, affinity='nearest_neighbors')
      models.fit(fimps)
      labelsx = models.fit_predict(fimps)
      #print(labelsx)


In [ ]:
prf = PrunedRandomForest(10, 100)
prf.fit(X, y)
print('KMEANS-')
for cluster_idx in prf.clusters:
          print(str(cluster_idx) + " ==> ", prf.clusters[cluster_idx])

prf1 = PrunedRandomForest()
prf1.fit(X, y)
print('DBSCAN-')
for cluster_idx in prf1.clusters:
          print(str(cluster_idx) + " ==> ", prf1.clusters[cluster_idx])

prf2 = PrunedRandomForest()
prf2.fit(X, y)
print('AGGLO')
for cluster_idx in prf2.clusters:
          print(str(cluster_idx) + " ==> ", prf2.clusters[cluster_idx])

prf3 = PrunedRandomForest()
prf3.fit(X,y)
print('Gaussian')
for cluster_idx in prf3.clusters:
          print(str(cluster_idx) + " ==> ", prf3.clusters[cluster_idx])

prf4 = PrunedRandomForest()
prf4.fit(X,y)
print('spectral clustering')
for cluster_idx in prf4.clusters:
          print(str(cluster_idx) + " ==> ", prf4.clusters[cluster_idx])


 fimps len = 100
Silhouetter Score: 0.001
 kmeans.labels_ len = 100
 fimps len = 100
 dbscan.labels_ len = 100
KMEANS-
1 ==>  [0, 6, 19, 21, 29, 35, 45, 48, 50, 70, 81, 83, 91]
5 ==>  [1, 13, 15, 17, 18, 27, 34, 36, 42, 61, 84, 97, 98, 99]
9 ==>  [2, 3, 14, 24, 28, 31, 44, 55, 69, 74, 75, 92, 93, 95]
2 ==>  [4, 30, 33, 39, 40, 54, 60]
8 ==>  [5, 7, 10, 11, 62, 66, 67, 72, 82, 85, 86, 89, 90]
4 ==>  [8, 26, 37, 38, 41, 47, 59, 63, 64, 73, 76, 77, 79, 80, 88, 94]
3 ==>  [9, 16, 43, 49, 53, 58, 65, 71]
6 ==>  [12, 20, 22, 25, 32, 51]
7 ==>  [23, 46, 52, 56, 57, 68, 78, 87]
0 ==>  [96]
 fimps len = 500
Silhouetter Score: 0.004
 kmeans.labels_ len = 500
 fimps len = 500
 dbscan.labels_ len = 500
DBSCAN-
6 ==>  [0, 6, 35, 43, 44, 61, 75, 78, 81, 97, 117, 119, 141, 147, 162, 165, 194, 196, 213, 218, 231, 237, 241, 265, 267, 276, 286, 295, 305, 325, 333, 335, 336, 351, 357, 358, 363, 366, 389, 394, 397, 401, 417, 429, 432, 439, 456, 467, 468, 475, 480, 485, 486, 487, 497]
9 ==>  [1, 3, 18, 25,

In [ ]:
prf.info

{0: {'accuracy': 0.5210721932695026, 'cluster_idx': 1},
 1: {'accuracy': 0.5265664767906919, 'cluster_idx': 5},
 2: {'accuracy': 0.5094735991596978, 'cluster_idx': 9},
 3: {'accuracy': 0.5139821435785561, 'cluster_idx': 9},
 4: {'accuracy': 0.5156789076071588, 'cluster_idx': 2},
 5: {'accuracy': 0.5183412111663233, 'cluster_idx': 8},
 6: {'accuracy': 0.5345614672968932, 'cluster_idx': 1},
 7: {'accuracy': 0.5129721649901022, 'cluster_idx': 8},
 8: {'accuracy': 0.5121157031470933, 'cluster_idx': 4},
 9: {'accuracy': 0.5265543570476307, 'cluster_idx': 3},
 10: {'accuracy': 0.515670827778451, 'cluster_idx': 8},
 11: {'accuracy': 0.530949783864582, 'cluster_idx': 8},
 12: {'accuracy': 0.49512382337494437, 'cluster_idx': 6},
 13: {'accuracy': 0.5112309619036076, 'cluster_idx': 5},
 14: {'accuracy': 0.5408960530036764, 'cluster_idx': 9},
 15: {'accuracy': 0.5076273583000039, 'cluster_idx': 5},
 16: {'accuracy': 0.5058659556417403, 'cluster_idx': 3},
 17: {'accuracy': 0.5112956005332687, 'clu

In [ ]:
# Highest acuracy selection-Cluster Representative
def getMaxAccuracyCluster(dicts):
    clusterObj = {}
    for key, value in dicts.items():
      if(clusterObj.get(str(value["cluster_idx"])) != None):
        clusterObj[str(value["cluster_idx"])].append({"index" : key, "accuracy" : value['accuracy']})
      else:
        clusterObj[str(value["cluster_idx"])] = []
        clusterObj[str(value["cluster_idx"])].append({"index" : key, "accuracy" : value['accuracy']})
    #Restructured the object so the key for dictionary is cluster_idx
    #Now you can access all the accuracies of cluster as below
    #print(clusterObj["2"])
    clusterMaxAccuracyObj = {}
    for clusterIndex in clusterObj.keys():
      max_acc = -1;
      for obj in clusterObj[clusterIndex]:
        if obj['accuracy'] == max_acc:
          clusterMaxAccuracyObj[str(clusterIndex)].append(obj['index'])
        if obj['accuracy'] > max_acc:
          max_acc = obj['accuracy']
          clusterMaxAccuracyObj[str(clusterIndex)] = []
          clusterMaxAccuracyObj[str(clusterIndex)].append(obj['index'])
    print(clusterMaxAccuracyObj)
    dict1 = {}
    print("Selecting first maximum accuracy of a decision tree in each cluster:")
    for key,val in clusterMaxAccuracyObj.items():
      output = {key : val[0]}
      print(output)
    return clusterMaxAccuracyObj

In [ ]:
getMaxAccuracyCluster(prf.info)

{'1': [70], '5': [27], '9': [14], '2': [39], '8': [86], '4': [94], '3': [71], '6': [51], '7': [46], '0': [96]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'1': 70}
{'5': 27}
{'9': 14}
{'2': 39}
{'8': 86}
{'4': 94}
{'3': 71}
{'6': 51}
{'7': 46}
{'0': 96}


{'1': [70],
 '5': [27],
 '9': [14],
 '2': [39],
 '8': [86],
 '4': [94],
 '3': [71],
 '6': [51],
 '7': [46],
 '0': [96]}

In [ ]:
def getTreeList(X, dicts):
    treeIdxs = []
    for trees in dicts.values():
        treeIdxs.append(trees[0])
    return treeIdxs

stored_ypred = []
def predict(X, treeList, prf):
    ypred = []
    for treeIdx in treeList:
        print(prf.decision_trees[treeIdx].predict(X))
        ypred.append(prf.decision_trees[treeIdx].predict(X))
    stored_ypred.append(ypred)
    print(type(ypred))
    return getMajorityLabels(ypred)

def getMajorityLabels(pred):
    predNp = np.array(pred)
    rows,cols = predNp.shape
    res = []
    for colIdx in range(cols):
        col = list(predNp[:, colIdx])
        res.append(max(set(col), key = col.count))
    return res




In [ ]:
treeList = getTreeList(X, getMaxAccuracyCluster(prf.info))

y_preed = predict(X_test, treeList, prf)

print("final answer = ", y_preed)
  
print("testing target data:",y_test)
  
print("Accuracy:",metrics.accuracy_score(y_test, y_preed))
print("Classification Report")
print(classification_report(y_test, y_preed))
print("precision", metrics.average_precision_score(y_test, y_preed))
print("recall", metrics.recall_score(y_test, y_preed))
print("f1", metrics.f1_score(y_test, y_preed))
print("f2", metrics.fbeta_score(y_test,y_preed, beta=2))

{'1': [70], '5': [27], '9': [14], '2': [39], '8': [86], '4': [94], '3': [71], '6': [51], '7': [46], '0': [96]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'1': 70}
{'5': 27}
{'9': 14}
{'2': 39}
{'8': 86}
{'4': 94}
{'3': 71}
{'6': 51}
{'7': 46}
{'0': 96}
[0. 1. 0. 1. 1. 1. 1. 0. 0. 1. 0. 1. 0. 1. 0. 0. 0. 0. 1. 1. 0. 1. 0. 0.
 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 0. 1.
 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 1. 0. 1. 0. 1. 1. 1. 1. 0. 1. 1. 0. 0.
 0. 0. 0. 1. 1. 0. 1. 1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 0. 0. 1. 0. 1.
 0. 0. 1. 1. 1. 0. 1. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0.
 1. 1. 0. 1. 0. 1. 1. 0. 1. 1. 0. 1. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 1. 0.
 1. 1. 0. 1. 1. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 1. 1. 1. 0. 1. 1. 0. 1.
 0. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 0. 0. 0. 0.
 1. 0. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0. 1. 1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1.
 0. 1. 1. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0. 0. 0. 1. 0. 1.

In [ ]:
treeList = getTreeList(X, getMaxAccuracyCluster(prf1.info))

y_preed = predict(X_test, treeList, prf1)

print("final answer = ", y_preed)
print(y_test)
  
print("Accuracy:",metrics.accuracy_score(y_test, y_preed))
print("Classification Report")
print(classification_report(y_test, y_preed))
print("precision", metrics.average_precision_score(y_test, y_preed))
print("recall", metrics.recall_score(y_test, y_preed))
print("f1", metrics.f1_score(y_test, y_preed))
print("f2", metrics.fbeta_score(y_test,y_preed, beta=2))

{'6': [351], '9': [217], '0': [139], '2': [410], '8': [207], '4': [19], '1': [361], '5': [388], '3': [42], '7': [339]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'6': 351}
{'9': 217}
{'0': 139}
{'2': 410}
{'8': 207}
{'4': 19}
{'1': 361}
{'5': 388}
{'3': 42}
{'7': 339}
[1. 1. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 1. 1. 1. 1. 0. 0.
 0. 0. 1. 1. 0. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0.
 0. 1. 1. 1. 1. 1. 0. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 1. 0. 1. 1. 0. 0.
 1. 0. 0. 0. 1. 0. 0. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1.
 0. 0. 1. 1. 0. 0. 1. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0.
 0. 0. 0. 1. 1. 1. 1. 0. 0. 1. 0. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.
 1. 1. 1. 0. 0. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1.
 0. 0. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0.
 1. 1. 0. 0. 1. 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 0. 0. 1. 0.
 0. 1. 1. 0. 0. 1. 0. 0. 1. 0. 0. 1. 1. 1. 1. 0

In [ ]:
treeList = getTreeList(X, getMaxAccuracyCluster(prf2.info))

y_preed = predict(X_test, treeList, prf2)

print("final answer = ", y_preed)
print(y_test)
  
print("Accuracy:",metrics.accuracy_score(y_test, y_preed))
print("Classification Report")
print(classification_report(y_test, y_preed))
print("precision", metrics.average_precision_score(y_test, y_preed))
print("recall", metrics.recall_score(y_test, y_preed))
print("f1", metrics.f1_score(y_test, y_preed))
print("f2", metrics.fbeta_score(y_test,y_preed, beta=2))

{'4': [35], '5': [160], '9': [79], '2': [231], '8': [256], '6': [438], '3': [489], '1': [395], '0': [372], '7': [451]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'4': 35}
{'5': 160}
{'9': 79}
{'2': 231}
{'8': 256}
{'6': 438}
{'3': 489}
{'1': 395}
{'0': 372}
{'7': 451}
[1. 0. 0. 1. 1. 0. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0.
 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 1. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1.
 0. 1. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 0. 0.
 0. 0. 0. 0. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1.
 0. 0. 0. 1. 0. 0. 1. 1. 0. 0. 1. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0.
 1. 1. 0. 1. 0. 0. 1. 1. 0. 1. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 1. 0.
 0. 1. 1. 1. 0. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 0. 1.
 0. 0. 1. 1. 1. 0. 1. 1. 1. 0. 0. 1. 0. 0. 1. 1. 0. 0. 0. 1. 1. 0. 0. 0.
 1. 0. 1. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 1. 0. 0. 1. 1. 1. 0. 0. 0. 1. 1.
 1. 1. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1

In [ ]:
treeList = getTreeList(X, getMaxAccuracyCluster(prf3.info))

y_preed = predict(X_test, treeList, prf3)

print("final answer = ", y_preed)
print(y_test)
  
print("Accuracy:",metrics.accuracy_score(y_test, y_preed))
print("Classification Report")
print(classification_report(y_test, y_preed))
print("precision", metrics.average_precision_score(y_test, y_preed))
print("recall", metrics.recall_score(y_test, y_preed))
print("f1", metrics.f1_score(y_test, y_preed))
print("f2", metrics.fbeta_score(y_test,y_preed, beta=2))

{'9': [446], '0': [33], '1': [142], '7': [44], '3': [79], '6': [475], '8': [136], '4': [24], '5': [287], '2': [119]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'9': 446}
{'0': 33}
{'1': 142}
{'7': 44}
{'3': 79}
{'6': 475}
{'8': 136}
{'4': 24}
{'5': 287}
{'2': 119}
[1. 1. 1. 0. 1. 0. 1. 1. 0. 1. 0. 1. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 0. 0.
 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 1. 1. 1. 1. 0. 1.
 0. 1. 1. 0. 1. 1. 0. 0. 1. 0. 0. 1. 0. 1. 0. 0. 1. 1. 1. 0. 0. 1. 0. 0.
 1. 0. 0. 0. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 1. 1. 1. 0. 0. 1. 0. 1.
 0. 0. 1. 0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 0. 1. 1. 1. 1.
 1. 1. 0. 1. 1. 1. 0. 0. 0. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 0.
 0. 1. 0. 1. 0. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1.
 0. 0. 0. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0.
 1. 1. 0. 0. 0. 0. 1. 1. 1. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 0. 1. 1. 1. 0.
 1. 1. 1. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 1. 1. 0.

In [ ]:
treeList = getTreeList(X, getMaxAccuracyCluster(prf4.info))

y_preed = predict(X_test, treeList, prf4)

print("final answer = ", y_preed)
print(y_test)
  
print("Accuracy:",metrics.accuracy_score(y_test, y_preed))
print("Classification Report")
print(classification_report(y_test, y_preed))
print("precision", metrics.average_precision_score(y_test, y_preed))
print("recall", metrics.recall_score(y_test, y_preed))
print("f1", metrics.f1_score(y_test, y_preed))
print("f2", metrics.fbeta_score(y_test,y_preed, beta=2))

{'3': [479], '9': [412], '1': [270], '5': [4], '0': [72], '8': [204], '2': [315], '6': [109], '4': [477], '7': [205]}
Selecting first maximum accuracy of a decision tree in each cluster:
{'3': 479}
{'9': 412}
{'1': 270}
{'5': 4}
{'0': 72}
{'8': 204}
{'2': 315}
{'6': 109}
{'4': 477}
{'7': 205}
[0. 1. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 1. 1. 1. 0. 0. 1. 1.
 1. 0. 1. 1. 1. 0. 0. 0. 1. 0. 1. 0. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1.
 0. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 0. 1.
 0. 0. 1. 0. 0. 0. 1. 1. 0. 0. 0. 1. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 0. 1.
 1. 0. 0. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 1. 0. 1. 1.
 1. 1. 0. 1. 0. 1. 0. 1. 0. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 0. 0. 0. 0. 1.
 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 0. 0. 0. 1. 1. 1. 0. 1. 0. 0.
 0. 0. 1. 1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1. 0. 0. 0. 0. 1.
 1. 1. 0. 0. 1. 0. 1. 1. 1. 0. 0. 0. 1. 1. 1. 0. 1. 1. 1. 1. 0. 0. 1. 1.
 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 1. 1. 1. 1. 1. 0. 

In [ ]:
#KERAS MODEL WITH TENSORFLOW- NEURAL NETWORKS

from numpy import loadtxt
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import confusion_matrix, accuracy_score 

#tensorflow
model_4 = tf.keras.Sequential([

                               tf.keras.layers.Dense(4, activation = 'relu'), #we may right it "tf.keras.activations.relu" too

                               tf.keras.layers.Dense(4, activation = 'relu'),

                               tf.keras.layers.Dense(1, activation = 'sigmoid')

])

model_4.compile( loss= tf.keras.losses.binary_crossentropy,

                optimizer = tf.keras.optimizers.Adam(lr = 0.01),

                metrics = ['accuracy'])

model_4.fit(X_train, y_train, epochs = 200, verbose = 2)

Epoch 1/200
25/25 - 5s - loss: 0.7408 - accuracy: 0.5225 - 5s/epoch - 191ms/step
Epoch 2/200
25/25 - 0s - loss: 0.6949 - accuracy: 0.5379 - 92ms/epoch - 4ms/step
Epoch 3/200
25/25 - 0s - loss: 0.6759 - accuracy: 0.5610 - 74ms/epoch - 3ms/step
Epoch 4/200
25/25 - 0s - loss: 0.6629 - accuracy: 0.5661 - 70ms/epoch - 3ms/step
Epoch 5/200
25/25 - 0s - loss: 0.6525 - accuracy: 0.5956 - 68ms/epoch - 3ms/step
Epoch 6/200
25/25 - 0s - loss: 0.6443 - accuracy: 0.6162 - 99ms/epoch - 4ms/step
Epoch 7/200
25/25 - 0s - loss: 0.6343 - accuracy: 0.6329 - 95ms/epoch - 4ms/step
Epoch 8/200
25/25 - 0s - loss: 0.6277 - accuracy: 0.6444 - 113ms/epoch - 5ms/step
Epoch 9/200
25/25 - 0s - loss: 0.6179 - accuracy: 0.6765 - 95ms/epoch - 4ms/step
Epoch 10/200
25/25 - 0s - loss: 0.6103 - accuracy: 0.6958 - 93ms/epoch - 4ms/step
Epoch 11/200
25/25 - 0s - loss: 0.6024 - accuracy: 0.7137 - 97ms/epoch - 4ms/step
Epoch 12/200
25/25 - 0s - loss: 0.5947 - accuracy: 0.7240 - 102ms/epoch - 4ms/step
Epoch 13/200
25/25 - 0s

In [ ]:
# predict probabilities for test set
from numpy import argmax
from sklearn.metrics import ( accuracy_score, roc_auc_score,
                             precision_score, recall_score, f1_score,fbeta_score,
                             classification_report
                             )
yhats = model_4.predict(X_test, verbose=0)
yhat = argmax(yhats, axis=-1).astype('int')
# evaluate on test set
yhat = model_4.predict(X_test)
yhat = argmax(yhat, axis=-1).astype('int')
acc = accuracy_score(y_test, yhat)
print('Accuracy: %.3f' % acc)

precision = precision_score(y_test,yhat)*100
recall = recall_score(y_test,yhat)*100
f1 = f1_score(y_test,yhat)*100

print("precision",precision)
print("recall",recall)
print("f1",f1)

print(classification_report(y_test, yhat))

11/11 [==============================] - 0s 2ms/step
Accuracy: 0.458
precision 0.0
recall 0.0
f1 0.0
              precision    recall  f1-score   support

           0       0.46      1.00      0.63       153
           1       0.00      0.00      0.00       181

    accuracy                           0.46       334
   macro avg       0.23      0.50      0.31       334
weighted avg       0.21      0.46      0.29       334



In [ ]:
n_features = X.shape[1]

n_class = len(unique(y))
# define the keras model
model = Sequential()
model.add(Dense(60, input_dim=n_features, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(30, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(n_class, activation='softmax'))
# compile the keras model
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam')
# fit the keras model on the dataset
model.fit(X_train, y_train, epochs=200, batch_size=32, verbose=2)
# evaluate on test set
yhat = model.predict(X_test)
yhat = argmax(yhat, axis=-1).astype('int')
acc = accuracy_score(y_test, yhat)
print('Accuracy: %.3f' % acc)

precision = precision_score(y_test,yhat)*100
recall = recall_score(y_test,yhat)*100
f1 = f1_score(y_test,yhat)*100

print("precision",precision)
print("recall",recall)
print("f1",f1)

print(classification_report(y_test, yhat))


Epoch 1/200
25/25 - 1s - loss: 0.8697 - 1s/epoch - 47ms/step
Epoch 2/200
25/25 - 0s - loss: 0.6070 - 59ms/epoch - 2ms/step
Epoch 3/200
25/25 - 0s - loss: 0.5143 - 57ms/epoch - 2ms/step
Epoch 4/200
25/25 - 0s - loss: 0.4523 - 58ms/epoch - 2ms/step
Epoch 5/200
25/25 - 0s - loss: 0.4137 - 73ms/epoch - 3ms/step
Epoch 6/200
25/25 - 0s - loss: 0.3495 - 57ms/epoch - 2ms/step
Epoch 7/200
25/25 - 0s - loss: 0.3037 - 63ms/epoch - 3ms/step
Epoch 8/200
25/25 - 0s - loss: 0.2659 - 55ms/epoch - 2ms/step
Epoch 9/200
25/25 - 0s - loss: 0.2300 - 55ms/epoch - 2ms/step
Epoch 10/200
25/25 - 0s - loss: 0.2022 - 61ms/epoch - 2ms/step
Epoch 11/200
25/25 - 0s - loss: 0.1923 - 62ms/epoch - 2ms/step
Epoch 12/200
25/25 - 0s - loss: 0.1548 - 57ms/epoch - 2ms/step
Epoch 13/200
25/25 - 0s - loss: 0.1387 - 61ms/epoch - 2ms/step
Epoch 14/200
25/25 - 0s - loss: 0.1120 - 63ms/epoch - 3ms/step
Epoch 15/200
25/25 - 0s - loss: 0.0966 - 73ms/epoch - 3ms/step
Epoch 16/200
25/25 - 0s - loss: 0.0870 - 88ms/epoch - 4ms/step
Ep

In [ ]:

scaler = StandardScaler()
scaled = scaler.fit_transform(X)
print(scaled)
X = pd.DataFrame(scaled)
print(X)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=1)


[[-0.23427536 -0.87076632 -0.5519429  ...  0.36611581 -0.37239391
   0.04849428]
 [-0.08906344 -0.20490089 -0.67694081 ... -1.54908636 -1.74031722
  -0.13508253]
 [ 1.9249399   2.4343237  -0.021955   ... -1.6390264   1.56279418
   0.85594144]
 ...
 [-0.17009248 -0.2166922   0.63608194 ...  0.28764747 -0.23313941
   0.14057102]
 [ 0.00560889  0.28478825  0.5226109  ... -0.13862661 -1.12131413
  -0.25860244]
 [-0.72625274  0.47149966  0.32674036 ... -0.74807932  0.3333198
   0.43727823]]
           0         1         2         3         4         5         6    \
0    -0.234275 -0.870766 -0.551943  0.647007 -0.513894  1.881730 -0.936612   
1    -0.089063 -0.204901 -0.676941  1.648936 -0.050750  0.760996 -0.180061   
2     1.924940  2.434324 -0.021955  0.646565  1.970107 -0.443700 -2.000153   
3    -1.881790 -2.054625  0.176561  0.753801  2.235826  2.337784 -0.065538   
4    -0.459759  0.288134 -0.630831 -0.047019  0.421873 -0.511784  1.121777   
...        ...       ...       ...       

In [ ]:
import sklearn
from sklearn.metrics import f1_score
# Making the Neural Network Classifier
NN = MLPClassifier(solver='lbfgs', alpha=1e-5,hidden_layer_sizes=(10, 30), random_state=1)

# Training the model on the training data and labels
NN.fit(X_train, y_train)

# Testing the model i.e. predicting the labels of the test data.
y_pred = NN.predict(X_test)

# Evaluating the results of the model
accuracy = accuracy_score(y_test,y_pred)*100
confusion_mat = confusion_matrix(y_test,y_pred)

# Step 8
# Printing the Results
print("Accuracy for Neural Network is:",accuracy)
print("Confusion Matrix")
print(confusion_mat)

precision = precision_score(y_test,y_pred)*100
recall = recall_score(y_test,y_pred)*100
f1 = f1_score(y_test,y_pred)*100
f2 = fbeta_score(y_test,y_pred, beta=2)*100
print("precision",precision)
print("recall",recall)
print("f1",f1)
print("f2",f2)


print(classification_report(y_test, y_pred))

Accuracy for Neural Network is: 55.97826086956522
Confusion Matrix
[[ 98  74]
 [ 88 108]]
precision 59.34065934065934
recall 55.10204081632652
f1 57.14285714285714
f2 55.90062111801242
              precision    recall  f1-score   support

           0       0.53      0.57      0.55       172
           1       0.59      0.55      0.57       196

    accuracy                           0.56       368
   macro avg       0.56      0.56      0.56       368
weighted avg       0.56      0.56      0.56       368

